dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.dependencies;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.aether.artifact.Artifact;
+
+import com.google.common.base.Joiner;
+
+/**
+ * Artifacts are considered to be the same if they have the same group ID,
+ * artifact ID, and version.
+ */
+final class DependencyPath {
+
+  private List<Artifact> path = new ArrayList<>();
+
+  void add(Artifact artifact) {
+    path.add(artifact);
+  }
+  
+  @Override
+  public String toString() {
+    return Joiner.on(" / ").join(path);
+  }
+  
+  @Override
+  public boolean equals(Object o) {
```

Comment:
Could we use equals/hashCode helper classes to reduce the boilerplate?

Reply:
I tried Objects.hash, but it didn't quite fit because I was hashing values from inside each object in a list and Artifact itself doesn't declare a hashCode method. I could build a list of hashable objects from the artifacts, and then hash that; but it ends up being about as complex as this. 

Objects.equals runs into a similar issue that Artifact does not have a reliable equals method that matches these semantics so we have to drop down further.